app/controllers/event24/TypeOfProtectionGroup1ReferenceController.scala

```diff
@@ -0,0 +1,115 @@
+import controllers.actions.{DataRetrievalAction, IdentifierAction}
+import forms.event24.TypeOfProtectionGroup1ReferenceFormProvider
+import models.enumeration.EventType
+import models.event24.{ProtectionReferenceData, TypeOfProtectionGroup1}
+import models.event24.TypeOfProtectionGroup1.SchemeSpecific
+import models.{Index, UserAnswers}
+import pages.Waypoints
+import pages.event24.{TypeOfProtectionGroup1Page, TypeOfProtectionGroup1ReferencePage}
+import play.api.data.FormError
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import uk.gov.hmrc.http.NotFoundException
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import views.html.event24.TypeOfProtectionGroup1ReferenceView
+
+import javax.inject.Inject
+import scala.annotation.tailrec
+import scala.concurrent.{ExecutionContext, Future}
+
+class TypeOfProtectionGroup1ReferenceController @Inject()(val controllerComponents: MessagesControllerComponents,
+                                                          identify: IdentifierAction,
+                                                          getData: DataRetrievalAction,
+                                                          userAnswersCacheConnector: UserAnswersCacheConnector,
+                                                          formProvider: TypeOfProtectionGroup1ReferenceFormProvider,
+                                                          view: TypeOfProtectionGroup1ReferenceView
+                                                   )(implicit ec: ExecutionContext) extends FrontendBaseController with I18nSupport {
+
+  private val form = formProvider()
+  private val eventType = EventType.Event24
+
+  def onPageLoad(waypoints: Waypoints, index: Index): Action[AnyContent] = (identify andThen getData(eventType)) { implicit request =>
+    val requiredReferenceTypes = getRequiredReferenceTypes(request.userAnswers, index)
+    val preparedForm = request.userAnswers.flatMap(_.get(TypeOfProtectionGroup1ReferencePage(index))).fold(form)(form.fill)
+    Ok(view(preparedForm, waypoints, index, requiredReferenceTypes))
+  }
+
+  def onSubmit(waypoints: Waypoints, index: Index): Action[AnyContent] = (identify andThen getData(eventType)).async {
+    implicit request =>
+      val requiredReferenceTypes = getRequiredReferenceTypes(request.userAnswers, index)
+      form.bindFromRequest().fold(
+        formWithErrors => {
+          val refTypesForm = form.bindFromRequest()
+
+          val validErrors = getValidErrors(refTypesForm.errors, Seq.empty, requiredReferenceTypes)
+
+          if (validErrors.nonEmpty) {
+            val updatedFormWithErrors = formWithErrors.copy(errors = validErrors)
+            Future.successful(BadRequest(view(updatedFormWithErrors, waypoints, index, requiredReferenceTypes)))
+          } else {
+              val formData = refTypesForm.data
+            val value = getUserAnswers(formData)
```

Comment:
can we pass **refTypesForm.data** to getUserAnswers() method, instead of local val formData